### PR TITLE
[Qwen3-TTS] Fuse no-compile Predictor kernels

### PR DESCRIPTION
**Pull request by junliu-mde**
## Motivation

Qwen3-TTS Predictor decode is launch-bound after the Talker compile path is
removed. The retained operations can run inside a bounded CUDA Graph while
preserving the fixed-seed output contract.

## Modifications

- Capture the per-token Predictor chain in bounded graphs keyed by batch bucket
  and sampling signature.
- Fuse exact seeded top-k/top-p sampling in the captured path.
- Fuse codec embedding gather/add inside graph capture.
- Use `torch.addmm` for the BF16, unquantized, TP1 H100 output-projection and
  residual path.
- Keep the eager PyTorch path for unsupported shapes, dtypes, devices, TP, and
  capture failures.
- Bind the bitonic parity implementation to the repository's `torch==2.13.0`
  pin and compare it directly with `torch.topk` in CUDA tests.
- Do not include the former KV-cache or first-token attention fusions. Both
  known-crash paths and their environment switches have been deleted.

The clean history separates the review boundaries: graph capture, seeded
sampling, output projection, sampling correctness fixes, and parity tests.

## Related Issues

Part of #1608. The KDA evaluation policy is tracked in #1650.

## Accuracy Test

The retained paths were previously exercised on H100 in 59 sampling tests and
72 Predictor/CUDA Graph tests. With the removed experimental fusions disabled,
the deterministic C4 and C8 closure matched original main exactly: 20/20 C4
WAVs and 40/40 C8 WAVs were byte-identical.

After rebuilding the branch on current main `3f819f9c`, the directly affected tests pass
on H100 with the pinned `torch==2.13.0`, `sglang==0.5.18`, and
`transformers==5.12.1` environment:

```text
121 passed in 96.53s
```

Repository GPU CI still requires a maintainer to add `run-ci` and
`run-qwen3-tts`.

## Benchmark & Profiling

The isolated seeded-sampling graph replay is 1.65x to 4.18x faster and uses 60
instead of 390 kernels. The latest retained integrated C4 profile records 109
Predictor graphs, 1,271 kernels per graph, and a 3.731537 ms median GPU-kernel
sum per graph.

These measurements isolate the Predictor path. This PR does not attribute the
full integrated end-to-end gain to one kernel.

## Checklist

- [x] Format the changed files with pre-commit.
- [x] Add graph, fallback, alias, endpoint, and seeded parity tests.
- [x] Delete known-crash experimental paths instead of shipping them disabled.
- [x] Rebuild the branch on current main without commits that introduce the
  removed paths.
- [ ] Run current-head GPU CI after a maintainer applies the required labels.

## CI

GPU CI requires a maintainer to add the `run-ci` and `run-qwen3-tts` labels.

### CODE REVIEW

File: tests/unit_test/qwen3_tts/test_predictor_kernels.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        token_ids,
+        embedding_weight,
+        gathered,
+        accumulated,
+    )
+    assert torch.equal(gathered, expected_gathered)
+    assert torch.equal(accumulated, expected_accumulated)
+
+
+@pytest.mark.skipif(not _HAS_CUDA, reason="Triton predictor kernel needs CUDA")
```
Comment (SandyLuXY) [inline tests/unit_test/qwen3_tts/test_predictor_kernels.py]: This CUDA test has no unconditional `accelerator` marker, so the repo's GPU selector `pytest -m accelerator` deselects this entire file. The two new fused embedding/output-projection graph tests are likewise unmarked.
Mark every CUDA-dependent case for accelerator CI while leaving the CPU fallback unmarked, and perform the CUDA availability check at runtime as required by `tests/README.md`.

File: sglang_omni/models/qwen3_tts/sampling_kernels.py [hunk 2]
```diff
@@ -87,8 +98,345 @@ def _seeded_gumbel_sample_sorted_kernel(
...
+        block_k: tl.constexpr,
+        has_top_p: tl.constexpr,
+    ):
+        row = tl.program_id(0)
+        vocab_offsets = tl.arange(0, 2048)
+        scores = tl.load(logits + row * logits_stride_b + vocab_offsets).to(tl.float32)
+        temperature = tl.maximum(tl.load(temperatures + row).to(tl.float32), 1e-5)
+        scores = scores / temperature
+
+        # PyTorch's top-k gather chooses lower input indices at its threshold.
```
Comment (SandyLuXY) [inline sglang_omni/models/qwen3_tts/sampling_kernels.py]: please follow comment style: `# Note (xxx): `

File: sglang_omni/models/qwen3_tts/sampling_kernels.py [hunk 3]
```diff
@@ -87,8 +98,345 @@ def _seeded_gumbel_sample_sorted_kernel(
...
+            high_bit_rank,
+            all_ones_rank,
+        )
+        sorted_scores = top_score_bits.to(tl.float32, bitcast=True)
+        sorted_token_ids = all_ones_rank - (
+            top_packed & all_ones_rank.to(tl.uint64)
+        ).to(tl.uint32)
+
+        if max_top_k <= 32:
+            # gatherTopK first writes every score strictly above the threshold
```
Comment (SandyLuXY) [inline sglang_omni/models/qwen3_tts/sampling_kernels.py]: same here

File: sglang_omni/models/qwen3_tts/sampling_kernels.py [hunk 4]
```diff
@@ -87,8 +98,345 @@ def _seeded_gumbel_sample_sorted_kernel(
...
+            # in source-index order. It then appends threshold-equal scores in
+            # source-index order. Its following 32-entry bitonic sort is
+            # unstable, so reproducing only the final top-k membership is not
+            # sufficient for seeded sampling.
+            threshold_ordered_score_bits = tl.max(
+                tl.where(ranks == max_top_k - 1, ordered_top_score_bits, 0),
+                axis=0,
+            )
+            # CUDA's threshold gather compares the float rank representation.
```
Comment (SandyLuXY) [inline sglang_omni/models/qwen3_tts/sampling_kernels.py]: same here

File: sglang_omni/models/qwen3_tts/sampling_kernels.py [hunk 5]
```diff
@@ -87,8 +98,345 @@ def _seeded_gumbel_sample_sorted_kernel(
...
+            equal_to_threshold = ordered_score_bits == threshold_ordered_score_bits
+            gather_order_keys = (
+                tl.where(
+                    greater_than_threshold,
+                    0,
+                    tl.where(equal_to_threshold, 1, all_ones_vocab),
+                ).to(tl.uint64)
+                << 32
+            ) | vocab_offsets.to(tl.uint64)
+            # ``tl.topk`` only supports descending selection. Complementing
```
Comment (SandyLuXY) [inline sglang_omni/models/qwen3_tts/sampling_kernels.py]: same here

File: sglang_omni/models/qwen3_tts/sglang_model.py [hunk 6]
```diff
@@ -1498,16 +1518,41 @@ def _sample_subtalker_token_seeded(
...
+        top_ps = self._sub_top_p_tensor.index_select(0, row_indices)
+        seeds = self._sub_sampling_seed_tensor.index_select(0, row_indices)
+        sub_positions = (
+            semantic_positions.to(device=logits.device, dtype=torch.long)
+            * max(int(self.config.num_code_groups) - 1, 1)
+            + int(layer_idx)
+            + 1
+        )
+
+        # The raw-logit fusion has a favorable launch-count tradeoff only in
```
Comment (SandyLuXY) [inline sglang_omni/models/qwen3_tts/sglang_model.py]: same here

File: tests/unit_test/qwen3_tts/test_predictor_kernels.py
Comment (SandyLuXY) [inline tests/unit_test/qwen3_tts/test_predictor_kernels.py:1]: Please add `test_predictor_kernels.py` and `test_sampling_kernels.py` to the test tree in `tests/README.md`

**Comment by SandyLuXY** [comment]
Adding E2E SeedTTS benchmark results for ref:

**Setup:**

- Original PR head: `82b9ffe5fe1eac32641c82e892d7e2397a19e81b`.
- Baseline: `main @ 3f819f9cdae3d4eeec22f73306c9067a1ec2542e`.
- Candidate: local no-push rebase of the PR onto that exact main, ` @9a0034a124c1564321b95aef714c6c01e60a15d7`.
- Model: `Qwen/Qwen3-TTS-12Hz-1.7B-Base`, BF16, unquantized, TP1.
- Dataset: `zhaochenyang20/seed-tts-eval-arrow`, complete 1088-sample English split.
- Capacity flags: `max_running_requests=64`, `max_queued_requests=64`, `cuda_graph_max_bs=64`.
- Hardware: NVIDIA H200 141 GB GPU

The candidate server log records Predictor CUDA Graph captures for sampled top-k 50 buckets 1, 2, 4, 8, 12, 16, 24, and 32, confirming that this E2E run exercised the optimized path.

### Throughput

| Concurrency | Main throughput (req/s) | Candidate throughput (req/s) | Delta | Main audio throughput (audio s/s) | Candidate audio throughput (audio s/s) | Delta |
| ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| 4 | 4.774 | 5.041 | **+5.59%** | 19.710 | 20.921 | **+6.14%** |
| 16 | 11.132 | 11.292 | **+1.44%** | 45.954 | 46.866 | **+1.98%** |
| 32 | 14.428 | 14.595 | **+1.16%** | 59.449 | 60.313 | **+1.45%** |

### Latency and RTF

| Concurrency | Main mean latency (s) | Candidate mean latency (s) | Delta | Main p95 latency (s) | Candidate p95 latency (s) | Delta | Main mean RTF | Candidate mean RTF | Delta |
| ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| 4 | 0.837 | 0.792 | **-5.38%** | 1.178 | 1.128 | **-4.24%** | 0.2089 | 0.1965 | **-5.94%** |
| 16 | 1.430 | 1.411 | **-1.33%** | 2.046 | 1.989 | **-2.79%** | 0.3523 | 0.3470 | **-1.50%** |
| 32 | 2.198 | 2.179 | **-0.86%** | 3.237 | 3.126 | **-3.43%** | 0.5423 | 0.5357 | **-1.22%** |

**Review by SandyLuXY — CHANGES_REQUESTED** [review]

File: tests/unit_test/qwen3_tts/test_predictor_kernels.py (hunk 1 above)
Comment (junliu-mde) [reply tests/unit_test/qwen3_tts/test_predictor_kernels.py]: Fixed in `1cd74abf`. Every CUDA-dependent case in `test_predictor_kernels.py` now has an unconditional `accelerator` marker, while the CPU fallback remains unmarked. The graph tests use a runtime fixture instead of collection-time CUDA probes, and the two fused tests are marked. `pytest --collect-only -m accelerator` selects 55 cases; the unmarked CPU subset passes 7 tests.

File: sglang_omni/models/qwen3_tts/sampling_kernels.py (hunk 5 above)
Comment (junliu-mde) [reply sglang_omni/models/qwen3_tts/sampling_kernels.py]: Fixed in `1cd74abf`. The comment now follows the `# Note (Jun Liu): ...` style.

File: sglang_omni/models/qwen3_tts/sglang_model.py (hunk 6 above)
Comment (junliu-mde) [reply sglang_omni/models/qwen3_tts/sglang_model.py]: Fixed in `1cd74abf`. The comment now follows the `# Note (Jun Liu): ...` style.

File: sglang_omni/models/qwen3_tts/sampling_kernels.py (hunk 3 above)
Comment (junliu-mde) [reply sglang_omni/models/qwen3_tts/sampling_kernels.py]: Fixed in `1cd74abf`. The comment now follows the `# Note (Jun Liu): ...` style.

File: sglang_omni/models/qwen3_tts/sampling_kernels.py (hunk 2 above)
Comment (junliu-mde) [reply sglang_omni/models/qwen3_tts/sampling_kernels.py]: Fixed in `1cd74abf`. The comment now follows the `# Note (Jun Liu): ...` style.

File: sglang_omni/models/qwen3_tts/sampling_kernels.py (hunk 4 above)
Comment (junliu-mde) [reply sglang_omni/models/qwen3_tts/sampling_kernels.py]: Fixed in `1cd74abf`. The comment now follows the `# Note (Jun Liu): ...` style.

File: tests/unit_test/qwen3_tts/test_predictor_kernels.py
Comment (junliu-mde) [reply tests/unit_test/qwen3_tts/test_predictor_kernels.py:1]: Fixed in `1cd74abf`. Both `test_predictor_kernels.py` and `test_sampling_kernels.py` are now listed in the Qwen3-TTS test tree.

**Comment by junliu-mde** [comment]
Addressed all review threads in `1cd74abf` and rebased onto current main `3f819f9c`. The accelerator selector now collects 55 CUDA cases while the unmarked CPU subset passes 7 tests. The three directly affected files pass on H100: `121 passed in 96.53s` with the repository-pinned environment. All review threads are resolved. @SandyLuXY please re-review. Repository GPU CI still needs `run-ci` and `run-qwen3-tts`.

**Review by SandyLuXY — APPROVED** [review]
lgtm
